### PR TITLE
Error handling in createWorkspace

### PR DESCRIPTION
**Pull request by blrubenstein**

### CODE REVIEW

File: api/src/main/java/org/pmiops/workbench/api/WorkspacesController.java [hunk 1]
```diff
@@ -320,8 +320,13 @@ private boolean checkWorkspaceWriteAccess(String workspaceNamespace, String work
               workspaceId.getWorkspaceName(),
               e.getResponseBody()),
           e);
-      // TODO: figure out what happens if the workspace already exists
-      throw new ServerErrorException("Error creating FC workspace", e);
+      if (e.getCode() == 403) {
```
Comment (danqrodney) [inline api/src/main/java/org/pmiops/workbench/api/WorkspacesController.java]: One thing missing from this: we have logic in generateFirecloudWorkspaceId that generates an FC name from a human readable name. It's possible for the human-readable names to differ while the generated names collide. 

I think probably what we should do is 1) have uniqueness checking of human-readable names in here, and throw a 409 if the name matches; and 2) update generateFirecloudWorkspaceId() so that it tacks on some numeric suffix or something to generate another FC name for the workspace when there's an FC name conflict, as we do for billing projects here:

https://github.com/all-of-us/workbench/blob/master/api/src/main/java/org/pmiops/workbench/api/ProfileController.java#L133

Basically, the user should just have to worry about making the display names unique; they shouldn't have to care about what the workspace name is in FC much.

Comment (danqrodney) [reply api/src/main/java/org/pmiops/workbench/api/WorkspacesController.java]: Somewhat related: we should probably enforce uniqueness of display names with a unique index in the database, and stop making description on workspaces be required. I had a PR to do that which I'll send out soonish.

**Comment by blrubenstein** [comment]
New version: Couple of things I think are worth bringing up:
1) This always appends a number to the end in FC, even if its the first with that FC name. Thoughts on that? Should I make one attempt before appending numbers?

2) Currently could get around the name collision by updating a workspaces name after creation. I intend to do a follow up PR to prevent that from happening.

File: api/src/main/java/org/pmiops/workbench/api/WorkspacesController.java [hunk 2]
```diff
@@ -301,33 +324,32 @@ private boolean checkWorkspaceWriteAccess(String workspaceNamespace, String work
...
-          String.format(
-              "Error creating FC workspace %s/%s: %s",
-              workspaceId.getWorkspaceNamespace(),
-              workspaceId.getWorkspaceName(),
-              e.getResponseBody()),
-          e);
-      // TODO: figure out what happens if the workspace already exists
-      throw new ServerErrorException("Error creating FC workspace", e);
+    FirecloudWorkspaceId attemptId = new FirecloudWorkspaceId(workspaceId.getWorkspaceNamespace(), workspaceId.getWorkspaceName());
+    for (int attemptValue = 1; attemptValue < 6; attemptValue++) {
```
Comment (danqrodney) [inline api/src/main/java/org/pmiops/workbench/api/WorkspacesController.java]: Nit: start at 0, and make 6 be a constant (e.g. MAX_ATTEMPT_VALUES.)

File: api/src/main/java/org/pmiops/workbench/api/WorkspacesController.java [hunk 3]
```diff
@@ -301,33 +324,32 @@ private boolean checkWorkspaceWriteAccess(String workspaceNamespace, String work
...
-          Level.SEVERE,
-          String.format(
-              "Error creating FC workspace %s/%s: %s",
-              workspaceId.getWorkspaceNamespace(),
-              workspaceId.getWorkspaceName(),
-              e.getResponseBody()),
-          e);
-      // TODO: figure out what happens if the workspace already exists
-      throw new ServerErrorException("Error creating FC workspace", e);
+    FirecloudWorkspaceId attemptId = new FirecloudWorkspaceId(workspaceId.getWorkspaceNamespace(), workspaceId.getWorkspaceName());
```
Comment (danqrodney) [inline api/src/main/java/org/pmiops/workbench/api/WorkspacesController.java]: Rename to fcWorkspaceId?

File: api/src/main/java/org/pmiops/workbench/api/WorkspacesController.java [hunk 4]
```diff
@@ -301,33 +324,32 @@ private boolean checkWorkspaceWriteAccess(String workspaceNamespace, String work
...
+    for (int attemptValue = 1; attemptValue < 6; attemptValue++) {
+      attemptId =
+          new FirecloudWorkspaceId(workspaceId.getWorkspaceNamespace(),
+              workspaceId.getWorkspaceName() + Integer.toString(attemptValue));
+      try {
+        attemptFirecloudWorkspaceCreation(attemptId);
+        break;
+      } catch (RuntimeException e) {
```
Comment (danqrodney) [inline api/src/main/java/org/pmiops/workbench/api/WorkspacesController.java]: Catch ConflictException here instead, and remove the instanceof below?

File: api/src/main/java/org/pmiops/workbench/api/WorkspacesController.java [hunk 5]
```diff
@@ -301,33 +324,32 @@ private boolean checkWorkspaceWriteAccess(String workspaceNamespace, String work
...
-              "Error creating FC workspace %s/%s: %s",
-              workspaceId.getWorkspaceNamespace(),
-              workspaceId.getWorkspaceName(),
-              e.getResponseBody()),
-          e);
-      // TODO: figure out what happens if the workspace already exists
-      throw new ServerErrorException("Error creating FC workspace", e);
+    FirecloudWorkspaceId attemptId = new FirecloudWorkspaceId(workspaceId.getWorkspaceNamespace(), workspaceId.getWorkspaceName());
+    for (int attemptValue = 1; attemptValue < 6; attemptValue++) {
+      attemptId =
```
Comment (danqrodney) [inline api/src/main/java/org/pmiops/workbench/api/WorkspacesController.java]: I would put this in the catch block. That way you don't assign attemptId above just to reassign it, and in the normal case there's no numeric suffix.

File: api/src/main/java/org/pmiops/workbench/api/WorkspacesController.java [hunk 6]
```diff
@@ -301,33 +324,32 @@ private boolean checkWorkspaceWriteAccess(String workspaceNamespace, String work
     }
     FirecloudWorkspaceId workspaceId = generateFirecloudWorkspaceId(workspace.getNamespace(),
```
Comment (danqrodney) [inline api/src/main/java/org/pmiops/workbench/api/WorkspacesController.java]: You can move this to after the if below. 

Another alternative is to have this method call get() on FC repeatedly until it finds a workspace name that's not taken (rather than trying to create one N times and handling ConflictException.)

I could see going either way.

**Comment by blrubenstein** [comment]
Just pushed a new version! This one should address all PR feedback, and fixes the tests! @danqrodney when you get a chance, PTAL.
